Swarm based backup

agazso opened the pull request:
Changes proposed in this pull request:
- Backups are stored encrypted on Swarm with 256 bit random encryption key
- Then a backup link is created that contains the Swarm hash and the encryption key
- This link can be optionally encrypted with a password
- This link can be sent over any messaging app, email etc. because it's short

Also added `opaqueTypes` as an experiment and separated the symmetric crypto functionality to the `cryptoHelpers` file.

vujevits commented on src/swarm/Swarm.ts:
```diff
@@ -144,14 +144,61 @@ const uploadData = async (data: string, swarmGateway: string): Promise<string> =
 
+const fetchArrayBuffer = (url: string, timeout: number): Promise<Uint8Array> => {
+    return new Promise((resolve, reject) => {
+        const request = new XMLHttpRequest();
```
why are we using here XMLHttpRequest and not fetch?

agazso replied:
React Native did not implement fetch with ArrayBuffer type and they consider this a feature request (which they don't want to implement) instead of a bug:

https://github.com/facebook/react-native/issues/9947

I found the workaround here:
https://stackoverflow.com/questions/43711977/how-to-use-fetch-api-with-arraybuffer-response-type-and-react-native

agazso commented on src/components/Restore.tsx:
```diff
@@ -5,38 +5,54 @@ import { SimpleTextInput } from './SimpleTextInput';
-        secretText: '',
-        canRestore: false,
+        backupPassword: '',
+        backupData: '' as HexString,
```
That would not work, because `backupData` is a `HexString`, which is a `BrandedString` and that always requires a typecast when creating from the string literal type.

agazso replied:
We could make `HexString` as a `FlavoredString` which has the definition of `string & { __tag__?: T}`, that is with the `__tag__` being optional, but then it also means everywhere you expect a `FlavoredString` you can pass in a regular string as well.

I chose to make it `BrandedString` because it gives you stronger type checks and that's the price you have to pay for it. I don't mind this, but if you don't like it we can always make a constructor function `makeHexString: (s: string) => HexString`

vujevits replied:
it's not a big deal :-)
But it's still a good practice to be explicit about default state type

agazso replied:
Done

agazso commented on src/components/Backup.tsx:
```diff
@@ -90,24 +117,64 @@ export class Backup extends React.PureComponent<Props, State> {
         }
     }
+
+    private backupAppStateToSwarm = async (secretHex: HexString): Promise<HexString> => {
```
I would not put these kind of loading state information in AppState unless it has to work across several screens. Instead it would be better to store this in a state variable, either as a class member or with hooks.

agazso commented on src/components/Restore.tsx:
```diff
@@ -54,49 +70,75 @@ export class Restore extends React.PureComponent<Props, State> {
-        if (!isValid) {
-            Alert.alert('Invalid backup!');
-            return;
+    private getDecryptedBackupData = (password: string): HexString => {
```
We have a file called `BackupRestore` and it belongs there. I will also put that file under the `helpers` dir.

agazso commented on src/components/Backup.tsx:
```diff
@@ -90,24 +117,64 @@ export class Backup extends React.PureComponent<Props, State> {
+
+    private calculateEncryptedBackupData = async (): Promise<HexString> => {
+        const backupData = this.getBackupData();
+        if (this.state.backupPassword !== '') {
```
We have a file called `BackupRestore` and it belongs there. I will also put that file under the `helpers` dir.

agazso commented on src/cryptoHelpers.ts:
```diff
@@ -0,0 +1,27 @@
+import nacl from 'ecma-nacl';
```
Done

agazso commented on src/opaqueTypes.ts:
```diff
@@ -0,0 +1,4 @@
+export type BrandedType<T, N>  = T & { __tag__: N };
```
Done

agazso commented:
There is one modification to how the backup works:
- If you don't enter a password, then the empty password (the hash of it) will be used as password, so there is no separate case for encrypted and non-encrypted backup links.

Other than that I moved the suggested files (and some more) under the `helpers` dir and cleaned up the logic. There is still some async code left in the `Backup.tsx` and `Restore.tsx` but I think it is good enough for now, considering also that these are only debug features.
